cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,247 @@
+package ext
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/bash"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/powershell"
+)
+
+// Generic action function that may return an error
+type ActionFn func() error
+
+// The type of command hooks. Supported values are 'pre' and 'post'
+type HookType string
+
+const (
+	// Executes pre-command hooks
+	HookTypePre HookType = "pre"
+	// Execute post-command hooks
+	HookTypePost HookType = "post"
+)
+
+// CommandHooks enable support to invoke integration scripts before & after commands
```

Comment:
@wbreza Should we try to decouple the "lifecycle phases" away from "command names"?

For example, as you alluded to, `preinfracreate` would only be triggered in `azd infra create`, but not in `azd provision`. This seems counterintuitive.

Having a well-defined list of phases also avoids the act of simply renaming a command to immediately break extensions. Also, this will work better for `prepackage` and `postpackage` which `package` is not a command.

Reply:
Interesting idea, however that would require more manual integration. Today the command hooks can run as a middleware which applies to any azd command.  If we only selectively support lifecycle phases manual updates would need to be required within each command/action we wanted to enable.

Going to keep for now and think about how to address commands with aliases like provision/infra create.